pallets/foreign-investments/src/lib.rs

```diff
@@ -0,0 +1,340 @@
+	/// which the investment originated. The lifecycle must go through the
+	/// following stages:
+	/// 	1. Increase redemption --> Initialize storage
+	/// 	2. Fully process pending redemption
+	/// 	3. Collect redemption
+	/// 	4. Trigger swap from pool to return currency
+	/// 	5. Completely fulfill swap order
+	/// 	6. Transfer back to source domain --> Kill storage entry
+	#[pallet::storage]
+	pub(super) type RedemptionState<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::AccountId,
+		Blake2_128Concat,
+		T::InvestmentId,
+		RedeemState<T::Balance, T::CurrencyId>,
+	>;
+
+	/// Maps `TokenSwapOrders` to `InvestmentInfo` to implicitly enable mapping
+	/// to `InvestmentState`.
+	///
+	/// NOTE: The storage is immediately killed when the swap order is
+	/// completely fulfilled even if the investment might not be fully
+	/// processed.
+	#[pallet::storage]
+	pub(super) type ForeignInvestmentInfo<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::TokenSwapOrderId, ForeignInvestmentInfoOf<T>>;
+
+	/// Maps an investor and their `InvestmentId` to the corresponding
+	/// `TokenSwapOrderId`.
+	///
+	/// NOTE: The storage is immediately killed when the swap order is
+	/// completely fulfilled even if the investment might not be fully
+	/// processed.
+	#[pallet::storage]
+	pub(super) type TokenSwapOrderIds<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::AccountId,
+		Blake2_128Concat,
+		T::InvestmentId,
+		T::TokenSwapOrderId,
+	>;
+
+	/// Maps an investor and their `InvestmentId` to the amount of collected
+	/// pool currency and the corresponding amount of tranche tokens burned for
+	/// the conversion based on the fulfillment price(s).
+	///
+	/// NOTE: The lifetime of this storage starts with collecting a redemption
+	/// in pool currency and ends with having swapped the entire amount to
+	/// return currency.
+	#[pallet::storage]
+	pub(super) type CollectedRedemptions<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::AccountId,
+		Blake2_128Concat,
+		T::InvestmentId,
+		CollectedInvestment<T::Balance>,
+		ValueQuery,
+	>;
+
+	/// Maps a `TokenSwapOrderId` to the corresponding `TokenSwapReason` for
+	/// which it was last updated, i.e. `Investment` or `Redemption`.
+	///
+	/// As there can always be at most a single active token swap for any
+	/// `TokenSwapOrderId`, and thus also for any `(AccountId, InvestmentId)`
+	/// pair, we only need to keep track of the last reason when we act upon a
+	/// notified status update for any ongoing swap. Otherwise, it would be
+	/// impossible to know whether an invest or a redeem transition needs to be
+	/// applied.
+	///
+	/// NOTE: The storage is immediately killed when the swap order is
+	/// completely fulfilled even if the investment might not be fully
+	/// processed.
+	#[pallet::storage]
+	pub(super) type TokenSwapReasons<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::TokenSwapOrderId, TokenSwapReason>;
```

Comment:
Could this storage be saved by adding `TokenSwapReason` as a value to `ForegignInvestmentInfo` storage?

Reply:
Great catch! `TokenSwapReasons` was added much later. Since both storages have the same lifetime and usage, they totally can be combined.